diff --git a/_inductor/runtime/triton_heuristics.py b/_inductor/runtime/triton_heuristics.py
--- a/_inductor/runtime/triton_heuristics.py
+++ b/_inductor/runtime/triton_heuristics.py
@@ -1677,7 +1677,7 @@
                 )
                 if hasattr(binary, "num_ctas")
                 else (
-                    (binary.metadata.num_ctas, *binary.metadata.cluster_dims)
+                    (getattr(binary.metadata, "num_ctas", 1), *getattr(binary.metadata, "cluster_dims", ()))
                     if hasattr(binary, "metadata")
                     else ()
                 )
